---
title: "scRNASeq data analysis-batch correction--MNN testing"
author: "Feng, Feng"
date: "`r Sys.Date()`"
header-includes:
    - \usepackage{setspace}\doublespacing
output: 
    html_document:
        dev: png
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = FALSE,
  comment = "#>",
  tidy=FALSE,
  engine='R'
)
```
# 1. Purpose
- Rerun and experiment with MNN testing different parameters

We will try mainly with different number of features to see how stable the method is.

# 2. Data input
The data input were generated by 10x Genomics Cellranger with all 6 subjects, 
and each subject with 3 cell populatoins. The data were on dropbox.

** Note **: to run this Rmd file, you need to revised the data
directory information in the following section.

```{r, eval=T}  
#Note: the original code were in file:///home/feng/Feng/hg/scravid2/CVIDagg6_SeuratPipeline_FF_v1.0.R
#library
library(Seurat)
library(dplyr)
library(Matrix)
library(gdata)
library(patchwork)
library(cowplot)
library(stringr)
library(ggpubr)
library(batchelor)
library(SeuratWrappers)
library(here)


  #load data
  #CVIDagg6.data <- Read10X(data.dir = "Documents/PhD Thesis Work/Data/Single Cell/CVIDagg6_05Dec2022/agg6_filtered_feature_bc_matrix")
CVIDagg6.data <- Read10X(data.dir = 
	"/home/feng/Windows/windowsD/feng/LAB/MSI/maglione/scRNA_analysis1_20220908/aggr6sub/aggr2212_6sbu/outs/count/filtered_feature_bc_matrix")
```

Now, we create Seurat object and clean up the data.

``` {r readData}
CVIDagg6 <- CreateSeuratObject(counts = CVIDagg6.data, project = "CVIDagg6", min.cells = 3, min.features = 200)
ct<-GetAssayData(CVIDagg6, slot="counts")
dat<-GetAssayData(CVIDagg6, slot="data")

#- Add meta.data
mito.genes <- grep(pattern = "^MT-", x = rownames(x = dat), value = TRUE)
percent.mito <- Matrix::colSums(dat[mito.genes, ]) / Matrix::colSums(dat)
CVIDagg6@meta.data$percent.mito <- percent.mito
CVIDagg6@meta.data$sub <- plyr::mapvalues(x = sapply(str_split(rownames(CVIDagg6@meta.data), "[-]"), function(x) x[2]), from = 1:18, to = rep(seq(1,6),c(3,3,3,3,3,3)))
CVIDagg6@meta.data$cvid <- plyr::mapvalues(x = sapply(str_split(rownames(CVIDagg6@meta.data), "[-]"), function(x) x[2]), from = 1:18, to = rep(c("HC","CVID","CVID","HC","CVID","HC"),c(3,3,3,3,3,3)))
CVIDagg6@meta.data$cells <- plyr::mapvalues(x = sapply(str_split(rownames(CVIDagg6@meta.data), "[-]"), function(x) x[2]), from = 1:18, to = rep(c("CD38_high","CD38_mid","CD38_low","CD38_high","CD38_mid","CD38_low","CD38_high","CD38_mid","CD38_low","CD38_high","CD38_low","CD38_mid","CD38_high","CD38_low","CD38_mid","CD38_high","CD38_low","CD38_mid"),c(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)))

#CVIDagg6@meta.data %>% head()

#- QC
CVIDagg6 <- subset(CVIDagg6, subset = nFeature_RNA > 200 & nCount_RNA < 40000 & percent.mito < .15)

CVIDagg6 
##An object of class Seurat 
##21188 features across 46878 samples within 1 assay 
##Active assay: RNA (21188 features, 0 variable features)


```
So far, we have done with initial steps. We will next run scaling on
uncorrected data to see whether there are batch effects.

Please DO NOT run the following section, since the code run very long 
time. In this case, we show the code and then jump to load previously 
saved data to do plotting.

```{r, eval=TRUE}
#normalize with default setting
CVIDagg6<-NormalizeData(CVIDagg6, normalization.method = "LogNormalize", scale.factor = 10000)

#find variable features
CVIDagg6 <- FindVariableFeatures(object = CVIDagg6, mean.function = ExpMean, dispersion.function = LogVMR, nfeatures = 20000)

CVIDagg6.orig<-CVIDagg6
#Idents(CVIDagg6.int) <- CVIDagg6.int@meta.data$seurat_clusters

```

We need to do it at cell population level.
```{r interByCells, eval=TRUE}
##############################
#	now integrate by cells ###
##############################

#subsetting cells by subject ID
ZA03.high<-subset(CVIDagg6, subset= sub==1&cells=="CD38_high" )
ZA03.mid<-subset(CVIDagg6, subset= sub==1&cells=="CD38_mid" )
ZA03.low<-subset(CVIDagg6, subset= sub==1&cells=="CD38_low" )
CC33.high<-subset(CVIDagg6, subset= sub==2&cells=="CD38_high")
CC33.mid<-subset(CVIDagg6, subset= sub==2&cells=="CD38_mid")
CC33.low<-subset(CVIDagg6, subset= sub==2&cells=="CD38_low")

CC22.high<-subset(CVIDagg6, subset= sub==3&cells=="CD38_high")
CC22.mid<-subset(CVIDagg6, subset= sub==3&cells=="CD38_mid")
CC22.low<-subset(CVIDagg6, subset= sub==3&cells=="CD38_low")

ZA61.high<-subset(CVIDagg6, subset= sub==4&cells=="CD38_high")
ZA61.mid<-subset(CVIDagg6, subset= sub==4&cells=="CD38_mid")
ZA61.low<-subset(CVIDagg6, subset= sub==4&cells=="CD38_low")

CC27.high<-subset(CVIDagg6, subset= sub==5&cells=="CD38_high")
CC27.mid<-subset(CVIDagg6, subset= sub==5&cells=="CD38_mid")
CC27.low<-subset(CVIDagg6, subset= sub==5&cells=="CD38_low")

ZA63.high<-subset(CVIDagg6, subset= sub==6&cells=="CD38_high")
ZA63.mid<-subset(CVIDagg6, subset= sub==6&cells=="CD38_mid")
ZA63.low<-subset(CVIDagg6, subset= sub==6&cells=="CD38_low")


```{r mnn2000, dpi=200, fig.retina=3, eval=F}
knitr::opts_chunk$set(dev.args = list(png = list(type = "cairo")))
###############################
###             MNN batch correction
#what to do next? need to do using scRNASeq and bachelor
#It can be used on seurat objects
#

CVIDagg6.int<-RunFastMNN(object.list=
	list(ZA03.high, ZA03.mid, ZA03.low, CC33.high,CC33.mid,CC33.low,
		CC22.high,CC22.mid,CC22.low, ZA61.high, ZA61.mid, ZA61.low,
		CC27.high, CC27.mid, CC27.low, ZA63.high, ZA63.mid, ZA63.low	
			)
	)
CVIDagg6.int <-RunUMAP(CVIDagg6.int, reduction = "mnn", 
	dims = 1:20, min.dist=0.75)
CVIDagg6.int <- FindNeighbors(CVIDagg6.int, reduction = "mnn", dims = 1:20)
CVIDagg6.int <- FindClusters(CVIDagg6.int,dims= 1:20, resolution = .6)
```

```{r mnn2000_b, dpi=200, fig.retina=3, eval=TRUE}
CVIDagg6.int<-readRDS(	
	file=here("Output","CVIDagg6_MNN_2K.rds"))
CVIDagg6.int
DefaultAssay(CVIDagg6.int)<-"mnn.reconstructed"

DimPlot(CVIDagg6.int, group.by = c("cvid", "cells", "sub","ident") )+theme(legend.text=element_text(size=rel(0.25)))

png(file="scRNACluster_MNN_clusters2K.png",
	width=1000, height=1000)

DimPlot(CVIDagg6.int, group.by = c("cvid", "cells", "sub","ident") )+theme(legend.text=element_text(size=rel(0.5)))
dev.off()

#saveRDS(file="CVIDagg6_MNN_2K.rds", CVIDagg6.int)

```



Now, let's do with different number of features(genes)

```{r mnn5000, eval=F}
###############################
###             MNN batch correction
#what to do next? need to do using scRNASeq and bachelor
#It can be used on seurat objects
#

CVIDagg6.int<-RunFastMNN(object.list=
	list(ZA03.high, ZA03.mid, ZA03.low, CC33.high,CC33.mid,CC33.low,
		CC22.high,CC22.mid,CC22.low, ZA61.high, ZA61.mid, ZA61.low,
		CC27.high, CC27.mid, CC27.low, ZA63.high, ZA63.mid, ZA63.low	
			), features=5000,
	)
CVIDagg6.int <-RunUMAP(CVIDagg6.int, reduction = "mnn", 
	dims = 1:20, min.dist=0.75)
CVIDagg6.int <- FindNeighbors(CVIDagg6.int, reduction = "mnn", dims = 1:20)
CVIDagg6.int <- FindClusters(CVIDagg6.int,dims= 1:20, resolution = .6)
```

```{r mnn5000_b, dpi=200, fig.retina=3, eval=TRUE}
CVIDagg6.int<-readRDS(	
	file=here("Output","CVIDagg6_MNN_5K.rds"))

DefaultAssay(CVIDagg6.int)<-"mnn.reconstructed"
print(CVIDagg6.int)

DimPlot(CVIDagg6.int, group.by = c("cvid", "cells", "sub","ident") )+theme(legend.text=element_text(size=rel(0.25)))

png(file="scRNACluster_MNN_clusters5K.png",
	width=1000, height=1000)

DimPlot(CVIDagg6.int, group.by = c("cvid", "cells", "sub","ident") )+theme(legend.text=element_text(size=rel(0.5)))
dev.off()

#saveRDS(file="CVIDagg6_MNN_5K.rds", CVIDagg6.int)
```
Try with 10000 genes
```{r mnn10000, eval=F}
###############################
###             MNN batch correction
#what to do next? need to do using scRNASeq and bachelor
#It can be used on seurat objects
#

CVIDagg6.int<-RunFastMNN(object.list=
	list(ZA03.high, ZA03.mid, ZA03.low, CC33.high,CC33.mid,CC33.low,
		CC22.high,CC22.mid,CC22.low, ZA61.high, ZA61.mid, ZA61.low,
		CC27.high, CC27.mid, CC27.low, ZA63.high, ZA63.mid, ZA63.low	
			),features=10000
	)
CVIDagg6.int <-RunUMAP(CVIDagg6.int, reduction = "mnn", 
	dims = 1:20, min.dist=0.75)
CVIDagg6.int <- FindNeighbors(CVIDagg6.int, reduction = "mnn", dims = 1:20)
CVIDagg6.int <- FindClusters(CVIDagg6.int,dims= 1:20, resolution = .6)
```

```{r mnn10000_b, dpi=200, fig.retina=3, eval=TRUE}
CVIDagg6.int<-readRDS(	
	file=here("Output","CVIDagg6_MNN_10K.rds"))

CVIDagg6.int
DimPlot(CVIDagg6.int, group.by = c("cvid", "cells", "sub","ident") )+theme(legend.text=element_text(size=rel(0.25)))
png(file="scRNACluster_MNN_clusters10K.png",
	width=1000, height=1000)

DimPlot(CVIDagg6.int, group.by = c("cvid", "cells", "sub","ident") )+theme(legend.text=element_text(size=rel(0.5)))
dev.off()

#saveRDS(file="CVIDagg6_MNN_10K.rds", CVIDagg6.int)

```
Last one, 20000 genes

No!!! the following is not run since it can not allocate enough memory for the matrix of 20000 features
```{r mnn20000, eval=F}
###############################
###             MNN batch correction
#what to do next? need to do using scRNASeq and bachelor
#It can be used on seurat objects
#

CVIDagg6.int<-RunFastMNN(object.list=
	list(ZA03.high, ZA03.mid, ZA03.low, CC33.high,CC33.mid,CC33.low,
		CC22.high,CC22.mid,CC22.low, ZA61.high, ZA61.mid, ZA61.low,
		CC27.high, CC27.mid, CC27.low, ZA63.high, ZA63.mid, ZA63.low	
			),features=20000
	)
CVIDagg6.int <-RunUMAP(CVIDagg6.int, reduction = "mnn", 
	dims = 1:20, min.dist=0.75)
CVIDagg6.int <- FindNeighbors(CVIDagg6.int, reduction = "mnn", dims = 1:20)
CVIDagg6.int <- FindClusters(CVIDagg6.int,dims= 1:20, resolution = .6)
```

```{r mnn20000_b, dpi=200, fig.retina=3, eval=TRUE}
CVIDagg6.int<-readRDS(	
	file=here("Output","CVIDagg6_MNN_20K.rds"))

CVIDagg6.int
DimPlot(CVIDagg6.int, group.by = c("cvid", "cells", "sub","ident") )+theme(legend.text=element_text(size=rel(0.25)))

png(file="scRNACluster_MNN_clusters20K.png",
	width=1000, height=1000)

DimPlot(CVIDagg6.int, group.by = c("cvid", "cells", "sub","ident") )+theme(legend.text=element_text(size=rel(0.5)))
dev.off()

#saveRDS(file="CVIDagg6_MNN_20K.rds", CVIDagg6.int)

```
DONE!!!!

```{r}
sessionInfo()
```

